File: haskell/glue/src/Glue/Lib/Math/Utility/Floor.hs
module Glue.Lib.Math.Utility.Floor where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Floor function
-- Mirrors Haskell Glue.Lib.Math.Utility.Floor.floor exactly
floor :: IR Eval
floor = NativeFunc floorImpl

-- Floor function implementation
-- Mirrors Haskell Glue.Lib.Math.Utility.Floor.floorImpl exactly
floorImpl :: IR Eval -> Eval (IR Eval)
floorImpl arg = case arg of
    Integer n -> pure $ Integer n
    Float n -> pure $ Integer (Prelude.floor n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/utility/floor.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Floor function (rounds down to nearest integer)
/// Mirrors Haskell Glue.Lib.Math.Utility.Floor.floor exactly
final Ir floor = IrNativeFunc(floorImpl);

/// Floor function implementation (rounds down to nearest integer)
/// Mirrors Haskell Glue.Lib.Math.Utility.Floor.floorImpl exactly
Eval<Ir> floorImpl(Ir arg) {
  return switch (arg) {
    IrInteger(value: final n) => Eval.pure(IrInteger(n)),
    IrFloat(value: final n) => Eval.pure(IrInteger(n.floor())),
    _ => throwError(wrongArgumentType(['number'])),
  };
}
